messages-nano.proto

```diff
@@ -0,0 +1,66 @@
+/*
+ * Messages (Nano specific) for KeepKey Communication
+ *
+ */
+
+syntax = "proto2";
+
+// Sugar for easier handling in Java
+option java_package = "com.keepkey.deviceprotocol";
+option java_outer_classname = "KeepKeyMessageNano";
+
+import "types.proto";
+
+/**
+ * Request: Ask device for Nano address corresponding to address_n path
+ * @next PassphraseRequest
+ * @next NanoAddress
+ * @next Failure
+ */
+message NanoGetAddress {
+    repeated uint32 address_n = 1;          // BIP-32 path to derive the key from master node
+    optional string coin_name = 2 [default='Nano'];
+    optional bool show_display = 3;         // optionally show on display before sending the result
+}
+
+/**
+ * Response: Contains a Nano address derived from device private seed
+ * @prev NanoGetAddress
+ */
+message NanoAddress {
+    required string address = 1;        // Coin address in Base58 encoding
+}
+
+/**
+ * Request: ask device to sign Nano transaction
+ * @start
+ * @next NanoSignedTx
+ */
+message NanoSignTx {
+    repeated uint32 address_n = 1;  // BIP-32 path to derive the key from master node
+    optional string coin_name = 2 [default='Nano'];
+    optional ParentBlock parent_block = 3;
+    optional bytes link_hash = 4;
+    optional string link_recipient = 5;
+    repeated uint32 link_recipient_n = 6;         // BIP-32 path to derive address for recipient
+    optional OutputAddressType address_type = 7;  // output address type
+    optional ExchangeType exchange_type = 8;      // exchange type data
+    required string representative = 9;
+    required bytes balance = 10;  // uint128 (big-endian encoded)
```

Comment:
Please don't use `required`. That constraint needs to be enforced by firmware instead of by the protobuf library. That affords us more flexibility if we need to change these message fields later.

Reply:
Got it. Will add explicit checks for those fields in the firmware and also integration tests to check that empty values are caught.

Reply:
I was just in the middle of doing that :) see: https://github.com/keepkey/device-protocol/pull/21/commits/98bcbf72eee4d4f65634c8f87403ed1552e8da89

Reply:
Other repos have been updated to handle the `required` -> `optional` changes.